src/pages/api/twitter-request-token.ts

```diff
@@ -0,0 +1,94 @@
+import { createHmac, randomBytes } from "crypto"
+import { NextApiHandler } from "next"
+
+const CONSUMER_KEY = process.env.TWITTER_API_KEY
+const CONSUMER_SECRET = process.env.TWITTER_API_SECRET
+const SIGNATURE_METHOD = "HMAC-SHA1"
+const OAUTH_VERSION = "1.0"
+const DEFAULT_CALLBACK_URL = "https://guild.xyz/oauth"
+const BASE_URL = "https://api.twitter.com/oauth/request_token"
+const KEY = `${encodeURIComponent(CONSUMER_SECRET)}&`
+const PARSE_FAIL_ERROR_MSG = "Unexpected data received from Twitter"
+
+// /oauth_token=(.*?)&oauth_token_secret=(.*?)&oauth_callback_confirmed=true/
+const parseV1Response = (response: string) => {
+  try {
+    const [, tokenStripped] = response.split("oauth_token=")
+    const [oauthToken, secretStripped] = tokenStripped.split("&oauth_token_secret=")
+    const [oauthTokenSecret] = secretStripped.split("&oauth_callback_confirmed=true")
+
+    if (!!oauthToken && !!oauthTokenSecret) {
+      return { oauthToken, oauthTokenSecret }
+    }
+  } catch (error) {
+    throw Error(PARSE_FAIL_ERROR_MSG)
+  }
+
+  throw Error(PARSE_FAIL_ERROR_MSG)
+}
+
+const handler: NextApiHandler = async (req, res) => {
+  if (req.method?.toLowerCase() !== "get") {
+    res
+      .status(400)
+      .json({ message: "Only GET requests are served by this endpoints" })
+    return
+  }
+
+  const { callbackUrl } = req.query
+  const oauthCallback = callbackUrl?.toString() ?? DEFAULT_CALLBACK_URL
+
+  const nonce = randomBytes(32).toString("base64")
+  const timestamp = Math.floor(Date.now() / 1000).toString()
+
+  const url = `${BASE_URL}?oauth_callback=${encodeURIComponent(oauthCallback)}`
+
+  const params = new URLSearchParams({
+    oauth_callback: oauthCallback,
+    oauth_consumer_key: CONSUMER_KEY,
+    oauth_nonce: nonce,
+    oauth_signature_method: SIGNATURE_METHOD,
+    oauth_timestamp: timestamp,
+    oauth_version: OAUTH_VERSION,
+  }).toString()
+
+  const signatureInput = `POST&${encodeURIComponent(BASE_URL)}&${encodeURIComponent(
+    params
+  )}`
+
+  const signature = createHmac("sha1", KEY)
+    .update(signatureInput)
+    .digest()
+    .toString("base64")
+
+  const authHeaderValue = `OAuth oauth_consumer_key="${encodeURIComponent(
+    CONSUMER_KEY
+  )}", oauth_nonce="${encodeURIComponent(
+    nonce
+  )}", oauth_signature="${encodeURIComponent(
+    signature
+  )}", oauth_signature_method="${SIGNATURE_METHOD}", oauth_timestamp="${timestamp}", oauth_version="${OAUTH_VERSION}"`
+
+  const response = await fetch(url, {
+    method: "POST",
+    headers: {
+      Authorization: authHeaderValue,
+    },
+  })
+
+  if (!response.ok) {
+    res.status(400).json({ message: "Failed to generate Twitter request token" })
```

Comment:
We could handle this error inside useOauthPopupWindow too, since right now I don't see an error message when the request fails, I can only see the Twitter oauth popup window, and an error inside that (it's because when we get this error when calling the serverless endpoint, we're sending an [object Object] in the oauth popup url as the oauth_token). ;)

Reply:
Nice catch!
Added error handling in https://github.com/agoraxyz/guild.xyz/pull/852/commits/69c0970e92412e430d480daa3830a9d9ff27abcd
And added a PostHog event about this error in https://github.com/agoraxyz/guild.xyz/pull/852/commits/a6f8d7bd4ac0c8375371604aa07c094c77219a8d since we can't see these errors from the BE logs